src/FixedVector/TH.hs

```diff
@@ -0,0 +1,70 @@
+{-# LANGUAGE TemplateHaskell #-}
+
+module FixedVector.TH where
+
+import FixedVector.Type
+import Language.Haskell.TH
+import Numeric.Natural (Natural)
+
+-- import Language.Haskell.TH.Lib
+
+-- create a function that we would have written as follows:
+-- createVector<N> :: t -> t -> {N times} -> t -> FixedVector t <N>
+-- createVector<N> v1 v2 {...} v<N> = FixedVector { _data = [ v1, v2, {...}, v<N> ] }
+-- using Template Haskell: https://hackage.haskell.org/package/template-haskell-2.19.0.0/docs/Language-Haskell-TH.html
+-- TODO: don't depend on internals of FixedVector and instead use exposed constructors
+--       this might only be possible once we have e.g. createVectorUpToN, because then we can depend on the existence of these constructors
+--       createVector<N> v1 v2 {...} v<N> = prepend v1 $ createVector<N-1> v2 ... v<N>
+createVectorN :: Natural -> DecsQ
+createVectorN n =
+  pure
+    [ SigD functionName functionType
+    , FunD functionName clauses
+    ]
+ where
+  -- here we have access to n from `createVector``
+  -- the name of the function we are creating
+  functionName :: Name
+  functionName = mkName $ "createVector" ++ show n
+
+  -- The type of elements stored in our vector, which we denote as `t`
+  vectorElementType :: Type
+  vectorElementType = VarT $ mkName "t"
+
+  -- ''FixedVector is the Name of the concrete type FixedVector
+  returnType :: Type
+  returnType = ConT ''FixedVector `AppT` vectorElementType `AppT` LitT (NumTyLit $ fromIntegral n)
+  -- alternatively, equivalent:
+  -- returnType = AppT (AppT (ConT ''FixedVector) vectorElementType) (LitT $ NumTyLit n)
+  -- returnType = foldr1 (AppT) [ConT ''FixedVector, vectorElementType, LitT (NumTyLit n)]
+
+  -- create the following expression:
+  -- t -> t -> ... -> t -> t -> FixedVector t n
+  -- in terms of our intermediate functions:
+  -- vectorElementType -> vectorElementType -> ... -> returnType
```

Comment:
```suggestion
  -- t -> t -> ... -> t -> t -> FixedVector t <N>
```

We made the implementation much more readable, so we can reduce the amount of explanations to what we expect to see as a result.